Test suite no longer passing with Python version 3.12

Dear aerleon maintainers,

I'd like to report an issue with the test suite that has to do with interpreter behavior changes in Python 3.12.x branch.

When using python 3.12.8 (I have not tested with earlier versions, but I believe the issue is the same), `arista_tp_test.AristaTpTest.testMissingMatchCriteria` fails with an error from the mocker:
`AttributeError: 'has_calls' is not a valid assertion. Use a spec for the mock if 'has_calls' is meant to be an attribute.`

Please see full test log below:
```sh
[lou@nixos:~/src/oss/aerleon]$ poetry run python --version && poetry run pytest
Python 3.12.8
======================================================= test session starts ========================================================
platform linux -- Python 3.12.8, pytest-7.4.4, pluggy-1.4.0
rootdir: /home/lou/src/oss/aerleon
collected 1978 items                                                                                                               

tests/api/api_test.py ...                                                                                                    [  0%]
tests/lib/aclcheck_test.py ....                                                                                              [  0%]
tests/lib/aclgenerator_test.py .........                                                                                     [  0%]
tests/lib/addressbook_test.py .....                                                                                          [  1%]
tests/lib/builtins_test.py ..................................................................                                [  4%]
tests/lib/fqdn_test.py .............                                                                                         [  5%]
tests/lib/nacaddr_test.py ...............                                                                                    [  5%]
tests/lib/naming_test.py .....................................................................................               [ 10%]
tests/lib/nsxt_test.py ..........                                                                                            [ 10%]
tests/lib/plugin_test.py ...                                                                                                 [ 10%]
tests/lib/policy_builder_test.py ..                                                                                          [ 10%]
tests/lib/policy_simple_test.py ......................                                                                       [ 11%]
tests/lib/policy_test.py ................................................................................................... [ 16%]
.......................                                                                                                      [ 18%]
tests/lib/summarizer_test.py ............                                                                                    [ 18%]
tests/lib/yaml_test.py .....................                                                                                 [ 19%]
tests/regression/arista/arista_test.py ...........                                                                           [ 20%]
tests/regression/arista_tp/arista_tp_test.py ....................................F.......................                    [ 23%]
tests/regression/aruba/aruba_test.py ...................................                                                     [ 25%]
tests/regression/brocade/brocade_test.py ....                                                                                [ 25%]
tests/regression/cisco/cisco_test.py .............................................                                           [ 27%]
tests/regression/ciscoasa/ciscoasa_test.py ......                                                                            [ 27%]
tests/regression/cisconx/cisconx_test.py ........                                                                            [ 28%]
tests/regression/ciscoxr/ciscoxr_test.py ............                                                                        [ 28%]
tests/regression/cloudarmor/cloudarmor_test.py ................                                                              [ 29%]
tests/regression/demo/demo_test.py ...                                                                                       [ 29%]
tests/regression/gce/gce_test.py ........................................................................................... [ 34%]
...................................................................                                                          [ 37%]
tests/regression/gcp/gcp_test.py ........................                                                                    [ 39%]
tests/regression/gcp_hf/gcp_hf_test.py ..............................................................................        [ 43%]
tests/regression/ipset/ipset_test.py ...........                                                                             [ 43%]
tests/regression/iptables/iptables_test.py ................................................................................. [ 47%]
.............................................                                                                                [ 50%]
tests/regression/juniper/juniper_test.py ................................................................................... [ 54%]
.................................................................................................................            [ 59%]
tests/regression/juniperevo/juniperevo_test.py .........................                                                     [ 61%]
tests/regression/junipermsmpc/junipermsmpc_test.py ......................................................................... [ 64%]
..................                                                                                                           [ 65%]
tests/regression/junipersrx/junipersrx_test.py ............................................................................. [ 69%]
...........................................................................                                                  [ 73%]
tests/regression/k8s/k8s_test.py .............................                                                               [ 74%]
tests/regression/nftables/nftables_test.py ........................................................                          [ 77%]
tests/regression/nokiasrl/nokiasrl_test.py ...................                                                               [ 78%]
tests/regression/nsxv/nsxv_test.py .....................................                                                     [ 80%]
tests/regression/openconfig/openconfig_test.py ...........                                                                   [ 81%]
tests/regression/packetfilter/packetfilter_test.py .........................................                                 [ 83%]
tests/regression/paloaltofw/paloaltofw_test.py .............................................                                 [ 85%]
tests/regression/pcap/pcap_test.py ...................                                                                       [ 86%]
tests/regression/sonic/sonic_test.py .                                                                                       [ 86%]
tests/regression/speedway/speedway_test.py ...                                                                               [ 86%]
tests/regression/srxlo/srxlo_test.py .........                                                                               [ 87%]
tests/regression/windows/windows_test.py ...                                                                                 [ 87%]
tests/regression/windows_advfirewall/windows_advfirewall_test.py ...........                                                 [ 87%]
tests/regression/windows_ipsec/windows_ipsec_test.py ........                                                                [ 88%]
tests/regression/yaml/yaml_regression_test.py ..........................................................................     [ 91%]
tests/regression_utils/capture_test.py ....                                                                                  [ 92%]
tests/unit/wrapwords_test.py ....                                                                                            [ 92%]
tests/utils/config_test.py .......                                                                                           [ 92%]
tests/utils/iputils_test.py ................................................................................................ [ 97%]
................................................                                                                             [100%]

============================================================= FAILURES =============================================================
______________________________________________ AristaTpTest.testMissingMatchCriteria _______________________________________________

self = <arista_tp_test.AristaTpTest testMethod=testMissingMatchCriteria>
mock_warn = <MagicMock name='warning' id='140114836602560'>

    @mock.patch.object(arista_tp.logging, "warning")
    @capture.stdout
    def testMissingMatchCriteria(self, mock_warn):
        atp = arista_tp.AristaTrafficPolicy(
            policy.ParsePolicy(GOOD_HEADER + MISSING_MATCH, self.naming), EXP_INFO
        )
        output = str(atp)
        self.assertNotIn("match", output, output)
>       mock_warn.has_calls(
            "WARNING: term %s has no valid match criteria and " "will not be rendered.",
            "missing-match",
        )

tests/regression/arista_tp/arista_tp_test.py:1005: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <MagicMock name='warning' id='140114836602560'>, name = 'has_calls'

    def __getattr__(self, name):
        if name in {'_mock_methods', '_mock_unsafe'}:
            raise AttributeError(name)
        elif self._mock_methods is not None:
            if name not in self._mock_methods or name in _all_magics:
                raise AttributeError("Mock object has no attribute %r" % name)
        elif _is_magic(name):
            raise AttributeError(name)
        if not self._mock_unsafe and (not self._mock_methods or name not in self._mock_methods):
            if name.startswith(('assert', 'assret', 'asert', 'aseert', 'assrt')) or name in _ATTRIB_DENY_LIST:
>               raise AttributeError(
                    f"{name!r} is not a valid assertion. Use a spec "
                    f"for the mock if {name!r} is meant to be an attribute.")
E               AttributeError: 'has_calls' is not a valid assertion. Use a spec for the mock if 'has_calls' is meant to be an attribute.

/nix/store/kjgslpdqchx1sm7a5h9xibi5rrqcqfnl-python3-3.12.8/lib/python3.12/unittest/mock.py:665: AttributeError
========================================================= warnings summary =========================================================
tests/utils/iputils_test.py:28
  /home/lou/src/oss/aerleon/tests/utils/iputils_test.py:28: PytestUnknownMarkWarning: Unknown pytest.mark.unit - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.unit

tests/api/api_test.py: 2 warnings
tests/lib/aclgenerator_test.py: 7 warnings
tests/lib/nsxt_test.py: 8 warnings
tests/regression/arista/arista_test.py: 11 warnings
tests/regression/arista_tp/arista_tp_test.py: 59 warnings
tests/regression/aruba/aruba_test.py: 35 warnings
tests/regression/brocade/brocade_test.py: 4 warnings
tests/regression/cisco/cisco_test.py: 50 warnings
tests/regression/ciscoasa/ciscoasa_test.py: 6 warnings
tests/regression/cisconx/cisconx_test.py: 8 warnings
tests/regression/ciscoxr/ciscoxr_test.py: 12 warnings
tests/regression/cloudarmor/cloudarmor_test.py: 16 warnings
tests/regression/demo/demo_test.py: 1 warning
tests/regression/gce/gce_test.py: 178 warnings
tests/regression/gcp_hf/gcp_hf_test.py: 65 warnings
tests/regression/ipset/ipset_test.py: 10 warnings
tests/regression/iptables/iptables_test.py: 122 warnings
tests/regression/juniper/juniper_test.py: 198 warnings
tests/regression/juniperevo/juniperevo_test.py: 25 warnings
tests/regression/junipermsmpc/junipermsmpc_test.py: 91 warnings
tests/regression/junipersrx/junipersrx_test.py: 160 warnings
tests/regression/k8s/k8s_test.py: 40 warnings
tests/regression/nftables/nftables_test.py: 20 warnings
tests/regression/nokiasrl/nokiasrl_test.py: 19 warnings
tests/regression/nsxv/nsxv_test.py: 33 warnings
tests/regression/openconfig/openconfig_test.py: 11 warnings
tests/regression/packetfilter/packetfilter_test.py: 45 warnings
tests/regression/paloaltofw/paloaltofw_test.py: 91 warnings
tests/regression/pcap/pcap_test.py: 21 warnings
tests/regression/sonic/sonic_test.py: 1 warning
tests/regression/speedway/speedway_test.py: 3 warnings
tests/regression/srxlo/srxlo_test.py: 9 warnings
tests/regression/windows/windows_test.py: 4 warnings
tests/regression/windows_advfirewall/windows_advfirewall_test.py: 11 warnings
tests/regression/windows_ipsec/windows_ipsec_test.py: 8 warnings
tests/regression/yaml/yaml_regression_test.py: 16 warnings
  /home/lou/src/oss/aerleon/aerleon/lib/aclgenerator.py:562: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
    current_date = datetime.datetime.utcnow().date()

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
===================================================== short test summary info ======================================================
FAILED tests/regression/arista_tp/arista_tp_test.py::AristaTpTest::testMissingMatchCriteria - AttributeError: 'has_calls' is not a valid assertion. Use a spec for the mock if 'has_calls' is meant to be an attribute.
========================================== 1 failed, 1977 passed, 1401 warnings in 12.45s ==========================================
```

--- a/tests/regression/arista_tp/arista_tp_test.py
+++ b/tests/regression/arista_tp/arista_tp_test.py
@@ -18,6 +18,7 @@
 import datetime
 import re
 from unittest import mock
+from unittest.mock import call
 
 from absl.testing import absltest
 
@@ -1002,9 +1003,9 @@ class AristaTpTest(absltest.TestCase):
         )
         output = str(atp)
         self.assertNotIn("match", output, output)
-        mock_warn.has_calls(
-            "WARNING: term %s has no valid match criteria and " "will not be rendered.",
-            "missing-match",
+        call_message = "WARNING: term %s has no valid match criteria and will not be rendered."
+        mock_warn.assert_has_calls(
+            [call(call_message, "missing-match"), call(call_message, "ipv6-missing-match")]
         )
         print(output)
 

